File: java/src/main/java/org/sammancoaching/TestStep.java
package org.sammancoaching;

import org.sammancoaching.dependencies.Config;
import org.sammancoaching.dependencies.Logger;
import org.sammancoaching.dependencies.Project;

public record TestStep(String name, Config config, Logger log) implements  PipelineStep {
    public PipelineStepResult run(Project project, PipelineStepResult previousStepResult) {
        boolean stepPassed;
        String failureReason = "";
        if (project.hasTests()) {
            if ("success".equals(project.runTests())) {
                log.info(name() + " passed");
                stepPassed = true;
            } else {
                String reason = name() + " failed";
                log.error(reason);
                stepPassed = false;
                failureReason = reason;
            }
        } else {
            log.info("No tests");
            stepPassed = true;
        }
        return new PipelineStepResult(name(), stepPassed, failureReason);
    }
}

File: csharp/BuildPipeline/TestStep.cs
namespace UntangledConditionals
{
    public class TestStep : PipelineStep
    {
        private readonly string _name;
        private readonly Config _config;
        private readonly Logger _log;

        public TestStep(string name, Config config, Logger log)
        {
            _name = name;
            _config = config;
            _log = log;
        }

        public PipelineStepResult Run(Project project, PipelineStepResult previousStepResult)
        {
            bool stepPassed;
            string failureReason = "";
            
            if (project.HasTests())
            {
                if ("success".Equals(project.RunTests()))
                {
                    _log.Info(_name + " passed");
                    stepPassed = true;
                }
                else
                {
                    string reason = _name + " failed";
                    _log.Error(reason);
                    stepPassed = false;
                    failureReason = reason;
                }
            }
            else
            {
                _log.Info("No tests");
                stepPassed = true;
            }

            return new PipelineStepResult(_name, stepPassed, failureReason);
        }
    }
}

